src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts

```diff
@@ -291,30 +292,22 @@ export class SettingsTreeSettingElement extends SettingsTreeElement {
 	}
 
 	matchesAnyFeature(featureFilters?: Set<string>): boolean {
-		const features = tocData.children!.filter(child => child.id === 'features')[0].children!.map(feature => {
-			return {
-				name: feature.id.substring(9),
-				values: feature.settings!.map(setting => setting.toString().substring(0, setting.toString().length - 2))
-			};
-		});
 
 		if (!featureFilters || !featureFilters.size) {
 			return true;
 		}
 
-		const result = Array.from(featureFilters).map(filter => {
-			const foundFeature = features.find(item => item.name === filter);
-			if (foundFeature) {
-				if (foundFeature.values.length === 1) {
-					return Array.from(featureFilters).some((feature: string) => this.setting.key.toLowerCase().startsWith(feature) && !this.setting.extensionInfo);
-				} else {
-					return Array.from(featureFilters).some((feature: string) => (this.setting.key.toLowerCase().startsWith(feature) || this.setting.key.toLowerCase().startsWith(foundFeature.values[1])) && !this.setting.extensionInfo);
-				}
+		const features = tocData.children!.find(child => child.id === 'features')!.children;
+
+		return Array.from(featureFilters).some(filter => {
+			const feature = features!.find(feature => 'features/' + filter === feature.id);
+			if (feature) {
+				const patterns = feature.settings?.map(setting => createSettingMatchRegExp(setting.toString()));
```

Comment:
`.toString()` is almost always suspicious in JS land because it rarely does what you want it to. Setting is already a string here right?

Reply:
setting is of type ISetting | string strangely

Reply:
would you prefer `as string`?

Reply:
I think we are using `ITOCEntry` for two things, the patterns and the resolved type. The cast is ok, you could straighten out that typing as well, like by giving `ITOCEntry` a generic type parameter